File: tests/poker.js
'use strict';

var chi = require('chi-squared');
var math = require('mathjs');

function getPokerTest(generator, options) {
	var observations = {};
    var observationType;

	for (var i = 0; i + 4 < options.sequenceLength; i += 5) {
        var valuesCount = {};
        var valueType;
        
        for (var j = 0; j < 5; j++) {
            valueType = generator.next();
            
            if (valuesCount[valueType] === undefined) {
                valuesCount[valueType] = 0;
            }
            
            valuesCount[valueType]++;        
        }
                
        var differentValuesCount = Object.keys(valuesCount).length;
                
        if (observations[differentValuesCount] === undefined) {
            observations[differentValuesCount] = 0;
        }
        
        observations[differentValuesCount]++
	}
	
    var tp = 1 / (options.valuesNumber * options.valuesNumber * options.valuesNumber * options.valuesNumber);
    var d = options.valuesNumber;
    
    var statistics = 0;
    
    var groupsNumber = Math.floor(options.sequenceLength / 5);
        
    for (var r = 0; r < 5; r++) {
        var p = tp * math.stirlingS2(5, r + 1);
        
        d--;
        tp *= d;
        
        if (observations[r + 1] === undefined) {
            observations[r + 1] = 0;
        }
        
        var expected = groupsNumber * p;
        
        var diff = (observations[r + 1] - expected);
        
        statistics +=  diff * diff  / expected;
    }
    
   	var pvalue = chi.cdf(statistics, 4);
	
	return {
        statistics: statistics,
        probability: 1 - pvalue,
        pvalue: pvalue
    };    
}

module.exports = getPokerTest;

File: tests/poker.py
import scipy.stats as stats
import math

def getPokerTest(generator, options):
    observations = {}
    observationType = None

    for i in xrange(0, options['sequenceLength'] - 4, 5):
        valuesCount = {}
        valueType = None

        for j in xrange(5):
            valueType = generator.next()

            if not valueType in valuesCount:
                valuesCount[valueType] = 0

            valuesCount[valueType] += 1

        differentValuesCount = len(valuesCount.keys())

        if not differentValuesCount in observations:
            observations[differentValuesCount] = 0

        observations[differentValuesCount] += 1

    tp = float(1) / math.pow(options['valuesNumber'], 4)
    d = options['valuesNumber']

    statistics = 0

    groupsNumber = math.floor(float(options['sequenceLength']) / 5)

    stirling2_5_i = [0, 1, 15, 25, 10, 1]

    for r in xrange(5):
        p = tp * stirling2_5_i[r + 1]

        d -= 1
        tp *= d

        if not (r + 1) in observations:
            observations[r + 1] = 0

        expected = groupsNumber * p;

        diff = observations[r + 1] - expected

        statistics +=  float(diff * diff)  / expected

    pvalue = 1 - stats.chi2.cdf(x=statistics, df=4)

    return {
        'statistics': statistics,
        'probability': 1 - pvalue,
        'pvalue': pvalue
    }
